lib/phoenix/socket/router.ex

```diff
@@ -0,0 +1,30 @@
+defmodule Phoenix.Socket.Router do
+  # Routes WebSocket and LongPoller requests.
+  @moduledoc false
+
+  use Plug.Builder
+  alias Phoenix.Transports.WebSocket
+  alias Phoenix.Transports.LongPoller
+
+  plug Plug.Logger
+  plug :fetch_query_params
+  plug :dispatch
+
+  def dispatch(conn, _) do
+    case {conn.method, transport(conn)} do
+      {"GET",  WebSocket}     -> WebSocket.call(conn, [])
+      {"POST", WebSocket}     -> WebSocket.call(conn, [])
+      {"OPTIONS", LongPoller} -> LongPoller.call(conn, :options)
+      {"GET", LongPoller}     -> LongPoller.call(conn, :poll)
+      {"POST", LongPoller}    -> LongPoller.call(conn, :publish)
+      _                       -> conn |> send_resp(:bad_request, "") |> halt()
+    end
+  end
+
+  defp transport(conn) do
+    case conn.query_params["transport"] do
```

Comment:
Ah, I see. In the endpoint, match on the ["ws"|_] prefix. This way we can continue to do route matching for long pooling instead of relying on query parameters.